syntax = "proto3";

service StatsService {
  rpc GetPostStats(PostIdRequest) returns (PostStatsResponse);
  rpc GetPostViewsOverTime(PostIdRequest) returns (TimeSeriesResponse);
  rpc GetPostLikesOverTime(PostIdRequest) returns (TimeSeriesResponse);
  rpc GetPostCommentsOverTime(PostIdRequest) returns (TimeSeriesResponse);
  rpc GetTopPosts(TopRequest) returns (TopPostsResponse);
  rpc GetTopUsers(TopRequest) returns (TopUsersResponse);
}

message PostIdRequest {
  string post_id = 1;
}

message TopRequest {
  string metric = 1; // "likes", "views", "comments"
}

message PostStatsResponse {
  int32 views = 1;
  int32 likes = 2;
  int32 comments = 3;
}

message TimeSeriesEntry {
  string date = 1;
  int32 count = 2;
}

message TimeSeriesResponse {
  repeated TimeSeriesEntry entries = 1;
}

message PostEntry {
  string post_id = 1;
  int32 count = 2;
}

message UserEntry {
  string user_id = 1;
  int32 count = 2;
}

message TopPostsResponse {
  repeated PostEntry posts = 1;
}

message TopUsersResponse {
  repeated UserEntry users = 1;
}
